### PR TITLE
Cosmosdb Add Role Definition and Role Assignment commands for Data Plane RBAC

### PR DESCRIPTION
**Pull request by nichatur**
Adding commands that allow public preview customers to create and manage Role Definitions and Role Assignments for enforcing data plane RBAC on Cosmos DB SQL accounts. The following commands are introduced:
- cosmosdb sql role definition create
- cosmosdb sql role definition delete
- cosmosdb sql role definition list
- cosmosdb sql role definition show
- cosmosdb sql role definition update
- cosmosdb sql role definition exists
- cosmosdb sql role assignment create
- cosmosdb sql role assignment delete
- cosmosdb sql role assignment list
- cosmosdb sql role assignment show
- cosmosdb sql role assignment update
- cosmosdb sql role assignment exists

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [x] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [x] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your PR is merged into master branch, a new PR will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repo and upgrade the version in the PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Cosmosdb

File: src/cosmosdb-preview/azext_cosmosdb_preview/_help.py [hunk 1]
```diff
@@ -145,3 +145,114 @@
 type: command
 short-summary: List all the databases and its collections that can be restored in the given account at the given timesamp and region.
 """
+
+helps['cosmosdb sql role'] = """
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_help.py:149]: Going forward we are attempting to make this docs as informative as possible. Please make sure to make these as elaborate and user friendly as possible

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py [hunk 2]
```diff
@@ -114,3 +121,19 @@ def load_arguments(self, _):
...
         c.argument('restore_location', options_list=['--restore-location', '-r'], help="The region of the restore.", required=True)
         c.argument('restore_timestamp_in_utc', options_list=['--restore-timestamp', '-t'], help="The timestamp of the restore", required=True)
+
+    account_name_type = CLIArgumentType(options_list=['--account-name', '-a'], help="Cosmosdb account name.")
+
+    # SQL role definition
+    with self.argument_context('cosmosdb sql role definition') as c:
+        c.argument('account_name', account_name_type, id_part=None)
+        c.argument('role_definition_id', options_list=['--id', '-i'], validator=validate_role_definition_id, help="Role Definition Id")
+        c.argument('role_definition_body', options_list=['--body', '-b'], validator=validate_role_definition_body, completer=FilesCompleter(), help="Role Definition body with Id (Optional for create), DataActions or Permissions, Type (Default is CustomRole), and AssignableScopes.  You can enter it as a string or as a file, e.g., --body @rdbody-file.json")
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:131]: nit: please add example of this body in the help message. You can refer IndexingPolicy in SQL container create command.

File: src/cosmosdb-preview/azext_cosmosdb_preview/_validators.py [hunk 3]
```diff
@@ -59,3 +61,117 @@ def validate_virtual_network_rules(ns):
...
+        if 'DataActions' in role_definition:
+            role_definition['Permissions'] = [Permission(data_actions=role_definition['DataActions'])]
+        else:
+            role_definition['Permissions'] = [Permission(data_actions=p['DataActions'])
+                                              for p in role_definition['Permissions']]
+
+        if 'Type' in role_definition:
+            role_definition['Type'] = role_definition['Type']
+        else:
+            role_definition['Type'] = "CustomRole"
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_validators.py:95]: Use the enum defined in for CustomRole Role type.

**Review by MehaKaushik — APPROVED** [review]
Minor comments, otherwise LGTM

File: src/cosmosdb-preview/azext_cosmosdb_preview/_help.py [hunk 4]
```diff
@@ -145,3 +145,114 @@
...
+examples:
+  - name: Create a SQL role definition under an Azure Cosmos DB account.
+    text: az cosmosdb sql role definition create -a MyAccount -g MyResourceGroup -b @role-definition.json
+"""
+
+helps['cosmosdb sql role definition delete'] = """
+type: command
+short-summary: Delete a SQL role definition under an Azure Cosmos DB account.
+examples:
+  - name: Create a SQL role definition under an Azure Cosmos DB account.
```
Comment (Juliehzl) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_help.py:171]: It should be `Delete` not `Create`

File: src/cosmosdb-preview/azext_cosmosdb_preview/_help.py [hunk 5]
```diff
@@ -145,3 +145,114 @@
...
+type: group
+short-summary: Manage Azure Cosmos DB SQL role definitions.
+"""
+
+helps['cosmosdb sql role definition create'] = """
+type: command
+short-summary: Create a SQL role definition under an Azure Cosmos DB account.
+examples:
+  - name: Create a SQL role definition under an Azure Cosmos DB account.
+    text: az cosmosdb sql role definition create -a MyAccount -g MyResourceGroup -b @role-definition.json
```
Comment (Juliehzl) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_help.py:164]: could you also provide examples for json string like this?
image

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py (hunk 2 above)
Comment (Juliehzl) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:131]: For parameter naming, I will recommend you to refer `az role definition` and `az role assign assignment` commands to be consistent. And customers could have similar usage experience.

Comment (nichatur) [reply src/cosmosdb-preview/azext_cosmosdb_preview/_params.py]: We already mimic the IAM Role definitions and Role assignments to a large degree, till the extent it is reasonable and understandable for the customer from a Cosmos DB standpoint.

**Review by Juliehzl — APPROVED** [review]

**Pull request merged**